tests/python/pants_test/scm/test_git.py

```diff
@@ -549,3 +554,148 @@ def test_detect_worktree_working_git(self):
         fp.write('echo ' + expected_worktree_dir)
       self.assertEqual(expected_worktree_dir, Git.detect_worktree())
       self.assertEqual(expected_worktree_dir, Git.detect_worktree(binary=git))
+
+
+class PreCommitHookTest(unittest.TestCase):
+
+  def setUp(self):
+    self.pants_repo_root = get_repo_root()
+
+  @contextmanager
+  def _create_tiny_git_repo(self):
+    with temporary_dir() as gitdir,\
+         temporary_dir() as worktree:
+      # A tiny little fake git repo we will set up. initialize_repo() requires at least one file.
+      readme_file = os.path.join(worktree, 'README')
+      touch(readme_file)
+      # The contextmanager interface is only necessary if an explicit gitdir is not provided.
+      with initialize_repo(worktree, gitdir=gitdir) as git:
+        yield git, worktree, gitdir
+
+  def _assert_subprocess_error(self, worktree, cmd, expected_excerpt):
+    with self.assertRaises(subprocess.CalledProcessError) as cm:
+      subprocess.check_output(cmd, cwd=worktree)
+    self.assertIn(expected_excerpt, str(cm.exception.output))
+
+  def _assert_subprocess_success(self, worktree, cmd, **kwargs):
+    self.assertEqual(0, subprocess.check_call(cmd, cwd=worktree, **kwargs))
+
+  def _assert_subprocess_success_with_output(self, worktree, cmd, full_expected_output):
+    output = subprocess.check_output(cmd, cwd=worktree)
+    self.assertEqual(full_expected_output, output)
+
+  def _assert_subprocess_error_with_input(self, worktree, cmd, input, expected_excerpt):
+    proc = subprocess.Popen(
+      cmd,
+      cwd=worktree,
+      stdin=subprocess.PIPE,
+      stdout=subprocess.PIPE,
+      stderr=subprocess.PIPE,
+    )
+    (stdout_data, stderr_data) = proc.communicate(input=input)
+    self.assertEqual(1, proc.returncode)
```

Comment:
For some reason I remember hearing that `self.assertNotEqual()` was a Bad Thing somewhere. I agree and have made the change.

Reply:
Yes, it is in general less good. Although there are still some times it’s valid to use. https://nedbatchelder.com/blog/201811/careful_with_negative_assertions.html

If we know the error code should be 1, I would keep it as 1. If there are multiple reasons it can fail, then we can do 

`self.assertIn(proc.returnCode, [1, 127])`

Reply:
I think since we are trying to [move to `set -e`](https://github.com/pantsbuild/pants/pull/7118/files#diff-996a98a3be9e08068af5a1ed85c8e731R25) which propagates the exit code of the failing command instead of using explicit `|| exit 1`, it would make sense to `self.assertNotEqual()` here.